support for copy dependencies

cherylking opened the pull request:
Fixes #705

scottkurz commented on docs/deploy.md:
```diff
@@ -9,12 +9,22 @@ The following are the parameters supported by this goal in addition to the [comm
 | --------  | ----------- | -------  |
 | appsDirectory | The server's `apps` or `dropins` directory where the application files should be copied. The default value is set to `apps` if the application is defined in the server configuration, otherwise it is set to `dropins`.  | No |
-| stripVersion | Strip artifact version when copying the application to Liberty runtime's application directory. The default value is `false`. | No |
+| copyDependencies | Copies the specified dependencies to the specified locations. Multiple `dependency` elements can be added to the `copyDependencies` configuration. The `location` element can be added to the `copyDependencies` or `dependency` configuration to override the default location, which is ${server.config.dir}/lib/global. | No |
```
`${server.config.dir}` is a server config property not a maven property... is this confusing at all ? Should we just spell it out what we mean instead of effectively referring to https://www.ibm.com/support/knowledgecenter/SS7K4U_liberty/com.ibm.websphere.wlp.zseries.doc/ae/rwlp_dirs.html doc ? (Or do we already do require the user to know this is what we're talking about?)

cherylking replied:
I'll reword it.

scottkurz commented on docs/deploy.md:
```diff
@@ -9,12 +9,22 @@ The following are the parameters supported by this goal in addition to the [comm
+| Parameter | Description | Required |
+| --------  | ----------- | -------  |
+| filter | The Maven coordinates `groupId:artifactId:version` identifying the dependency to be copied. The `artifactId` and `version` are optional. If only `groupId` is specified for the `filter`, all resolved dependencies with a matching `groupId` are copied to the specified or default location along with their transitive dependencies. If `groupId:artifactId` is specified for the `filter`, the r ...
+| location | The optional directory to which the dependency is copied. This can be an absolute path, or a relative path to the ${server.config.dir}. | No |
```
same comment on _server.config.dir_

cherylking replied:
I'll reword it.

scottkurz commented on liberty-maven-plugin/src/it/appsdirectory-notset-configured-it/src/test/java/net/wasdev/wlp/maven/test/app/DefaultAppDirectoryTest.java:
```diff
@@ -79,4 +79,22 @@ public void testApplicationFileExist() throws Exception {
+
+    @Test
+    public void testCopyDependenciesFilesExist() throws Exception {
+        File f = new File("liberty/usr/servers/test/lib/global/commons-logging-1.0.4.jar");
```
That path looks unusual.. isn't it usually:  `target/liberty/wlp/usr/servers...`  ?

cherylking replied:
It's relative to the particular test directory. I copied the path from the existing test method testApplicationFileExist() and I assure you it works. If I change the filename or path, it fails.

scottkurz replied:
OK, I think it's a different default when you use an           `<assemblyArtifact>` as the install source in config.

scottkurz commented on liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/applications/Dependency.java:
```diff
@@ -0,0 +1,57 @@
+     * Filter specifying the dependency to copy. The format is groupId:artifactId:version, with artifactId and 
+     * version optional. Maven dependency management will be used to resolve the dependency.
+     */
+    @Parameter(property = "filter")
```
This and location `@Parameter` doesn't look right but I'm fuzzy on this. If you did -Dfilter=xyz from the CLI how would maven know which dependency to parameterize?  Would it do all of them?  Or are you thinking it's helpful if you happen to just have 1?

cherylking replied:
I don't think supporting CLI properties for these makes sense. I'll change this to remove the property name.

scottkurz commented on liberty-maven-plugin/src/it/appsdirectory-notset-configured-it/pom.xml:
```diff
@@ -31,6 +31,11 @@
+    <dependency>
+        <groupId>org.apache.derby</groupId>
+        <artifactId>derby</artifactId>
+        <version>10.15.2.0</version>
```
Maybe make this 'test' scope to better cover:  `All scopes are included`

cherylking replied:
There is one that is test scope already..the commons-logging one.

scottkurz commented on liberty-maven-plugin/src/it/appsdirectory-notset-configured-it/pom.xml:
```diff
@@ -77,6 +82,22 @@
+                <filter>org.apache.derby:derbyclient:10.15.2.0</filter>
+            </dependency>
+          </copyDependencies>
+
```
Maybe add one more IT and cover the default `location` in `copyDependencies/location` ?  Doesn't seem covered anywhere.

cherylking replied:
You are correct that particular case is not covered. I could not cover all three scenarios in one IT, but I guess I can add another one.

scottkurz replied:
Thanks.

scottkurz commented on liberty-maven-plugin/src/main/java/io/openliberty/tools/maven/AbstractLibertySupport.java:
```diff
@@ -209,6 +210,113 @@ protected Artifact getArtifact(String groupId, String artifactId, String type, S
+            }
+        }
+
+        return resolvedDependencyFiles;
```
So what happens if there are conflicts?   E.g.   A->B->D.1  and A->C->D.2  ?     Are we following the algorithm Maven uses to resolve conflicts (e.g. https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html ) or are we relying on the first/last one read?   (I'm just going to ask instead of trying to read).

I'm concerned too because when Maven resolves dependency conflicts, IIUC, if you read the typical resolution algorithm it resolves conflicts _for a single scope_.   If we're trying to collect dependencies across multiple scopes, what happens if you have say A->B->D.1  in compile scope, and C->E->D.2  in test scope?

cherylking replied:
Honestly, I have no idea. This was the only way I could get any semblance of finding transitive dependencies for a particular dependency and am hoping that is good enough. The maven-dependency-plugin:copy-dependencies does not support copying transitive dependencies if either the include or exclude parameters are used to basically filter which dependencies to copy.

scottkurz replied:
Would the code copy both, or just the first (or last)?   If it copies both, can the filter be used to differentiate the one you want vs. the one you don't want?   I don't have a simple pre-existing example to test with or would mention it.    If we can't rationalize a behavior could we consider blowing up if we see two different versions somewhere?

cherylking replied:
I like the idea of blowing up if we see two different versions of the same dependency, but I think that would potentially make this unusable for certain scenarios. I don't see how the filter can be used to differentiate which one you want. When I call getArtifacts() on the project, it is supposed to return all dependencies and transitive dependencies needed for the project. Would it have already weeded out the conflicts? I think we need a concrete scenario to try and see if this current code "does the right thing".

scottkurz replied:
Here's a simple pair of dependencies leading to a conflict:  https://maven.apache.org/enforcer/enforcer-rules/dependencyConvergence.html  

So two common mechanisms for selecting one or the other would be:

```
    <dependencyManagement>
       <dependencies>
           <dependency>
             <groupId>org.slf4j</groupId>
             <artifactId>slf4j-api</artifactId>
             <version>1.6.1</version>
           </dependency>
        </dependencies>
    </dependencyManagement>
```

and

```
    <dependency>
      <groupId>org.slf4j</groupId>
      <artifactId>slf4j-nop</artifactId>
      <version>1.6.0</version>
      <exclusions>
        <exclusion>
            <groupId>org.slf4j</groupId>
            <artifactId>slf4j-api</artifactId>
        </exclusion>
    </exclusions>
    </dependency>
```
both selecting v 1.6.1.

It's a bit tricky since maven doesn't flag this situation as an error unless you add that dependencyConvergence enforcer rule... it'll pick one, I think the first?

So we might first stop and make sure that both the dependencyMgmt and exclusion approach work so that only one version of the dependency is seen, but if two are seen, IDK.. we could pick one.. is that the same as say `mvn dependency:tree` will show?  Not sure.

cherylking replied:
Ok..I tried adding the following two dependencies to create the conflict...and only one version of slf4j-api was seen and copied.

```    <dependency>
      <groupId>org.slf4j</groupId>
      <artifactId>slf4j-nop</artifactId>
      <version>1.6.0</version>
    </dependency>
```
```    <dependency>
      <groupId>org.slf4j</groupId>
      <artifactId>slf4j-jdk14</artifactId>
      <version>1.6.1</version>
    </dependency>
```

Whichever one is configured first, that is the transitive dependency the code sees. When sfl4j-nop is first, we see version 1.6.0 of slf4j-api. When slf4j-jdk14 is first, we see version 1.6.1 of slf4j-api.

Then I tried adding the `dependencyManagement` section you listed, and it got version 1.6.1 of slf4j-api even when slf4j-nop was listed first. And finally I tried using the `exclusion` instead of `dependencyManagement`, and once again it got version 1.6.1 of slf4j-api. So all in all good news I think.

cherylking commented:
@jjvilleg I added the stripVersion parameter and limited the scope to not include 'test' or 'provided' scoped dependencies. I also updated the documentation and the test cases.
